### PR TITLE
Feature: Add DetourFindRemotePayload and improve other payload-related methods

### LINKED ISSUE #79: Feature: DetourFindRemotePayload and DetourFindRemotePayloadEx for remote process.
In my use-case, it's possible that a DLL gets injected and uninjected multiple times in the target process's lifetime. I want to add a payload to the target process for my DLL to check if it is in the target process rather than the injecting process which uses SetWindowsHookEx to ensure that the DLL is uninjected by the system if my process dies because the detour is used to transmit information to my application, so it is not useful (or cannot live at all) if my application is not running. SetWindowsHookEx requires that I need to load the DLL in the injecting process.

Since it's possible for the DLL to be uninjected, calling DetourCreateRemotePayload each time it needs to be injected will pollute the process with multiple copies of the magic payload. I cannot use a static boolean variable to track if the remote payload exists, because the case where the DLL gets injected or uninjected is either my process or the target process starting or closing (my process restarting would reset that boolean variable)

Currently, I would need to copy DetourFindPayload's implementation to my own code as well as using CreateToolhelp32Snapshot (or use ReadProcessMemory and read the module headers) to know if the payload already exists in the target process.

This should be a feature of Detours and the signature of the two functions should be the same than the in-process ones, except the void pointer returned is suitable for further reading or writing via ReadProcessMemory or WriteProcessMemory. This would have the side effect of allowing some sort of live configuration system for the detour, where the detour can just read configuration from the payload and the injector can write to it to dynamically change the behavior of the detour instead of having to establish another method of communication with the injector process (which would have been more complex for both parties).

### PR DESCRIPTION
**Pull request by sylveon**
Other improvements are:
- Makes the `pcbData` parameter in DetourFindPayload and DetourFindPayloadEx optional, so that if an application only needs to search for the presence of a payload, they can ignore the size by passing `nullptr`.
- Makes the `pvData` parameter in DetourCopyPayloadToProcess const, so that a pointer to a const C++ object can be passed instead of the object needing to be `const_cast`ed or being non-const.
- Adds DetourCopyPayloadToProcessEx, which has the same interface than DetourCopyPayloadToProcess, but it returns the address of the payload in the remote module, if the program later wants to write to it.

Fixes #79

From basic testing, it seems to work for me. I can apply corrections and add a sample if required (although finding a trivial use case for those APIs would be required)

### CODE REVIEW

**Comment by sylveon** [comment]
@dtarditi could this and other PRs/issues be addressed? I'd like at some point to start using this in production (it's been working flawlessly in development)

**Comment by dtarditi** [comment]
@sylveon, what is the motivation for wanting to add DetourFindRemotePayload?   The other changes look straightforward.   It isn't clear to me how this new function would help with the core goal of detours, which is allow ppeople to instrument and monitor and API calls.

**Comment by sylveon** [comment]
My use-case is described in #79.

Basically, I use the payload as a flag to check if the DLL is actually loaded in the process I want to detour API calls (much like the current CreateProcessWithDll code does) and I may use it in the future to transmit information to the DLL when it is injected (like a shared handle or a pointer to a memory map).

However, the way my app works, the user can close and reopen it at will, so the DLL and detour may be removed (as in, completely unloaded from its address space) and added multiple times within the lifetime of the target process.

Without this API, I cannot know if a payload already exists in the target process, so I always have to create a new payload, making its potential future purpose of transmitting information useless since there is not guarantee about which payload DetourFindPayload will give you if there are multiple payloads with the same GUID, and it also leaks memory each time my app is closed and relaunched due to Detours not providing a way to deallocate the payload from the target process.

With this API, I can instead just find an existing payload (or create a new payload if there is none yet), not having to allocate a new one each time, and then edit the payload's content using WriteProcessMemory calls.

This function can very well be written standalone, but since it taps into implementation details of Detours, which might change in the future, I think that having it as part of the core codebase would be better.

**Comment by sylveon** [comment]
Note that my commits became unverified due to me changing my GPG key, you can ignore that.

**Comment by dtarditi** [comment]
@sylveon, this use case makes sense.  Could you add a test case for the new API function?

**Comment by sylveon** [comment]
@dtarditi I fixed a bug related to memory allocation (used `delete` instead of `delete[]`, fixed that issue by avoiding heap allocations entirely), and added a sample. The sample is not representative of my real world use, but it should be good enough to test and demonstrate the API. What it does is:
- creates a child process
- creates a payload in that child process containing a handle of the parent process
- parent process adds a payload to itself containing zeroes using `DetourCopyPayloadToProcessEx` and remembers the memory address of the payload data
- child process grabs the handle to the parent process using `DetourFindPayloadEx`
- child process finds the payload in the parent process using `DetourFindRemotePayload` and fills it with random data using `WriteProcessMemory`
- child process exits using the same random data as exit code
- parent process makes sure the payload data matches the child process exit code, and returns with 1 if it doesn't

File: src/creatwth.cpp
Comment (bgianfo) [inline src/creatwth.cpp:146]: Can you add SAL annotations to the new functions please? 
https://docs.microsoft.com/en-us/cpp/code-quality/annotating-function-parameters-and-return-values?view=vs-2019

File: src/modules.cpp
Comment (bgianfo) [inline src/modules.cpp:827]: Do you want to send these as separate PR's so we can merge them faster?

**Comment by bgianfo** [comment]
It you don't mind, it would be nice if you could include the markdown for the WIKI documentation pages for these new API's in this PR (just in the PR description, or as comments) we can just paste them in once we merge.

File: samples/payload/Makefile
Comment (bgianfo) [inline samples/payload/Makefile:12]: People routinely report and fix issues compiling detours with MSVC 2005 and other crazy old configurations, how do I'm worried about taking a dependency on modenish c++, I agree it would be nice but not sure how realistic it is.

File: samples/payload/payload.cpp
Comment (bgianfo) [inline samples/payload/payload.cpp:38]: winrt::check_bool throws, seems like we are missing a try/catch block to report failure?

Comment (sylveon) [reply samples/payload/payload.cpp:38]: If an uncaught exception is thrown, the process will automatically exit with an non-0 code. Since cppwinrt needs to go away (because it requires C++17) I'll look into adding some logging.

File: samples/payload/Makefile
Comment (sylveon) [reply samples/payload/Makefile:12]: I added that for cppwinrt but I'll remove it.

Comment (sylveon) [reply samples/payload/Makefile]: Should work with old compilers now

**Comment by bgianfo** [comment]
@dtarditi, do you think this needs more work? Or is it ready to merge?

File: src/modules.cpp
Comment (bgianfo) [inline src/modules.cpp:796]: I think this is going to get fixed in #139 FYI, and will lead to merge conflict.

File: src/detours.cpp
Comment (bgianfo) [inline src/detours.cpp:2550]: I realize  this is just a nice refactor of the existing code, but is there a reason not use IsEqualGUID that I didn't notice?

Comment (sylveon) [reply src/detours.cpp:2550]: Using `IsEqualGUID` makes detours depend on ole32.lib.

Comment (bgianfo) [reply src/detours.cpp:2550]: Uhh.. maybe they changed it in different versions of the headers...

In my windows SDK (version 10.0.19041.0) `IsEqualGUID` is declared as a macro or inline function depending on how you are compiling. 

```cpp

#if !defined (__midl)
#if !defined (_SYS_GUID_OPERATORS_)
#define _SYS_GUID_OPERATORS_
#include <string.h>

// Faster (but makes code fatter) inline version...use sparingly
#ifdef __cplusplus
__inline int InlineIsEqualGUID(REFGUID rguid1, REFGUID rguid2)
{
   return (
      ((unsigned long *) &rguid1)[0] == ((unsigned long *) &rguid2)[0] &&
      ((unsigned long *) &rguid1)[1] == ((unsigned long *) &rguid2)[1] &&
      ((unsigned long *) &rguid1)[2] == ((unsigned long *) &rguid2)[2] &&
      ((unsigned long *) &rguid1)[3] == ((unsigned long *) &rguid2)[3]);
}

__inline int IsEqualGUID(REFGUID rguid1, REFGUID rguid2)
{
    return !memcmp(&rguid1, &rguid2, sizeof(GUID));
}

#else   // ! __cplusplus

#define InlineIsEqualGUID(rguid1, rguid2)  \
        (((unsigned long *) rguid1)[0] == ((unsigned long *) rguid2)[0] &&   \
        ((unsigned long *) rguid1)[1] == ((unsigned long *) rguid2)[1] &&    \
        ((unsigned long *) rguid1)[2] == ((unsigned long *) rguid2)[2] &&    \
        ((unsigned long *) rguid1)[3] == ((unsigned long *) rguid2)[3])

#define IsEqualGUID(rguid1, rguid2) (!memcmp(rguid1, rguid2, sizeof(GUID)))

#endif  // __cplusplus
```

Comment (sylveon) [reply src/detours.cpp:2550]: Ah, I was going off the [documentation](https://docs.microsoft.com/en-us/windows/win32/api/guiddef/nf-guiddef-isequalguid)... We just need to know since when that was made inline. For example, I'm not sure if whatever SDK supported by VS 2005 also does that :/

Comment (bgianfo) [reply src/detours.cpp:2550]: Sure, it looks like it still exports IsEqualGUID... 
```
$ dumpbin /exports C:\Windows\System32\ole32.dll | rg IsEqualGUID
       1139  140 00010CA0 IsEqualGUID
```

I think your change is correct, maybe we should just add a comment to
DetourAreSameGuid, explaining that it exists for backwards compat.

File: src/modules.cpp
Comment (sylveon) [reply src/modules.cpp:796]: I'll locally merge once the other PR is merged

File: src/detours.cpp
Comment (sylveon) [reply src/detours.cpp:2556]: I added a comment

File: src/modules.cpp
Comment (sylveon) [reply src/modules.cpp]: Seems like this didn't even cause a merge conflict, since the code diff was the exact same.

**Comment by sylveon** [comment]
Any update on this?

**Comment by bgianfo** [comment]
@dtarditi waiting on your feedback here.

**Review by dtarditi — APPROVED** [review]
Looks good to me.

**Comment by bgianfo** [comment]
@sylveon I would like to include this in the next release can you: 
- [ ] Fix up the merge conflicts
- [ ] Add some basic unit test coverage of this in the new [unit tests](https://github.com/microsoft/Detours/tree/master/tests).

**Comment by sylveon** [comment]
Yeah I was actually planning to take a look at the conflicts tonight. I'll see for the unit tests as well.

**Comment by sylveon** [comment]
I resolved conflicts and added unit test @bgianfo 
While I was at it I noticed quite a few random tabs where present in source code, so I find & replace'd them, which will help in preparation to #181

**Comment by bgianfo** [comment]
Sorry can you submit the tab changes as a seperate PR? Since the repro is configured to only allow squash commits, that's going to make the commit diff pretty messy. ☹

**Comment by sylveon** [comment]
Reverted that commit for now

**Comment by sylveon** [comment]
There seems to be some contamination between PR builds, the 2019 builds are complaining about tests for DetourFindPayloadEx already existing in test_modules_api.cpp at line 623, but the file isn't even that long in my branch.

**Comment by sylveon** [comment]
Looks like I've got conflicts to resolve anyways - will do tonight

**Review by bgianfo — APPROVED** [review]

**Comment by bgianfo** [comment]
Thanks again for the contribution @sylveon!

**Comment by bgianfo** [comment]
I merged all of the doc changes that you provided in your gist as well. 
See: https://github.com/microsoft/Detours/wiki/_compare/6c0fef6d375e2eebf558be1f6c3c8a817a4121a4...da14a2afdadfa3c7a4018b7e44485fb2825ce00c

**Pull request merged**
